### PR TITLE
[ENH] add images_to_coordinates

### LINKED ISSUE #423: threshold z-map images from Dataset objects to create coordinates
It would be useful to create coordinates from an image within a Dataset object in circumstances where:
- for some of the studies you only have coordinates, but others you have images
- want to compare an IBMA approach with a CBMA approach when you have images.
This can be accomplished with nilearns [get_clusters_table](https://nilearn.github.io/modules/generated/nilearn.reporting.get_clusters_table.html)

### PR DESCRIPTION
**Pull request by jdkent**
Closes #423 .

Changes proposed in this pull request:

- use transformer class to make ImageToCoordinates
- test ImageToCoordinates
- add notebook demonstrating use of ImageToCoordinates
- add `z_to_p` function for cases when z-maps are not available, but you still specify `z_threshold` as a z-value
- test `z_to_p`
- fix bug in ` dict_to_coordinates`

- [x] How should I handle circular dependencies for `get_masker`?
  - have a round about somewhat hacky solution
- [x] how to handle missing required stat maps?
  - I decided to add a `merge_strategy` parameter with three different options for users to decide whether to overwrite existing coordinates, and to skip generating coordinates for study contrasts without the required z or p maps.
- [x] somewhat related to the above point, if I include peak statistics for the coordinates that are generated, how should the coordinates that already exist (and do not have peak statistics reported) be represented? As `NaN`s?
  - Representing missing peak coordinates as `NaN`s is my current solution.
- [x] how should negative and positive statistics be handled?
  - there is `two_sided` parameter now available thanks to @tsalo 
- [x] how should I represent metadata
  - have a `coordinate_source` column with either `original` or `nimare` (or None)

### CODE REVIEW

File: nimare/transforms.py
Comment (tsalo) [inline nimare/transforms.py:244]: It would be awesome if you could retain z-statistics and/or contrast values in the coordinate version of the dataset as well.

**Comment by jdkent** [comment]
I think this is ready for review, @tyarkoni  and @tsalo.
My current biggest question is how to work around the circular imports, I shuffled some functions around, but my solution may be non-optimal.
Once the docs test pass, I'll link to the example notebook I created demonstrating usage.

File: nimare/transforms.py
Comment (jdkent) [reply nimare/transforms.py]: done! (at least a first pass is done)

**Review by tyarkoni — COMMENTED** [review]
Looks good! Only a couple of minor comments, one of which is largely obsoleted by our conversation.

File: nimare/transforms.py
Comment (tsalo) [inline nimare/transforms.py:258]: I just added a "two_sided" flag to `nilearn.reporting.get_clusters_table`. Can/should that be leveraged here?

Comment (jdkent) [reply nimare/transforms.py:258]: Yes, that is perfect and I will use that, thanks!

File: examples/02_meta-analyses/plot_compare_ibma_and_cbma.py
Comment (tsalo) [inline examples/02_meta-analyses/plot_compare_ibma_and_cbma.py:69]:
```suggestion
# Compare CBMA and IBMA Z-maps
```

File: nimare/utils.py
Comment (tsalo) [inline nimare/utils.py:557]:
```suggestion
        LGR.info("Input is an ambiguous 3x3 matrix.\nAssuming coords are row vectors (Nx3).")
```

Good opportunity to fix this mistake introduced by `black`.

File: nimare/utils.py
Comment (tsalo) [inline nimare/utils.py:452]:
```suggestion
    description="Validates the Lancaster MNI-to-Talairach and Talairach-to-MNI transforms.",
```

File: nimare/utils.py
Comment (tsalo) [inline nimare/utils.py:484]:
```suggestion
        LGR.info("Input is an ambiguous 3x3 matrix.\nAssuming coords are row vectors (Nx3).")
```

File: nimare/utils.py
Comment (tsalo) [inline nimare/utils.py:527]:
```suggestion
    description="Validates the Lancaster MNI-to-Talairach and Talairach-to-MNI transforms.",
```

File: nimare/transforms.py
Comment (tsalo) [inline nimare/transforms.py:209]:
```suggestion
    merge_strategy : {'fill', 'replace', 'demolish'}, optional
```

Pretty sure this is the preferred numpydoc approach.

File: .circleci/config.yml
Comment (tsalo) [inline .circleci/config.yml:21]: You'll want to do that for all of the other instances of `deps9` in this file.

BTW I never actually thought to use that before! I normally make a dummy change to `setup.py`, which I now realize is a very silly approach.

Comment (jdkent) [reply .circleci/config.yml:21]: actually, I do not need to change that, I changed `nimare/info.py`, so that should actually update the cache, forgot to revert this change but glad you learned something from it.

File: nimare/info.py
Comment (jdkent) [inline nimare/info.py:57]: necessary to use `two_sided` from nilearn's create_cluster_table

File: nimare/transforms.py
Comment (jdkent) [reply nimare/transforms.py:326]: marking as resolved, went with 'fill', 'replace', and 'demolish' as merge strategies

**Comment by jdkent** [comment]
I think with one final look over, this should be ready to merge.

File: nimare/transforms.py
Comment (tsalo) [inline nimare/transforms.py:339]: At least for `fill` and `replace`, should the `CoordinateGenerator` also update the study metadata? Not sure if it should go in the coordinates DataFrame or the metadata one, but it seems like something that could be taken into account in some CBMAs.

File: nimare/utils.py
Comment (tsalo) [inline nimare/utils.py:92]: Is this a bug in the current version?

File: nimare/transforms.py
Comment (tsalo) [inline nimare/transforms.py:254]: It looks like `seed` is unused. Can it be dropped?

File: nimare/transforms.py
Comment (tsalo) [inline nimare/transforms.py:255]: I think this needs a more detailed docstring.

File: examples/01_datasets/transform_images_to_coordinates.py
Comment (tsalo) [inline examples/01_datasets/transform_images_to_coordinates.py:82]:
```suggestion
# * 'replace' will create coordinates for study contrasts with images.
```

File: examples/01_datasets/transform_images_to_coordinates.py
Comment (tsalo) [inline examples/01_datasets/transform_images_to_coordinates.py:138]:
```suggestion
# Run a meta analysis using each of the strategies.
```

File: examples/01_datasets/transform_images_to_coordinates.py
Comment (tsalo) [inline examples/01_datasets/transform_images_to_coordinates.py:140]:
```suggestion
# The difference is because in 'fill' most of the original coordinates
```

**Review by tsalo — CHANGES_REQUESTED** [review]
Everything is looking really good. Most of my comments are about docstrings, although there is one about whether metadata should be logged. Also, there is a conflict that needs to be addressed. Finally, can you remove `[WIP]` from the PR title since it's ready for review?

File: nimare/transforms.py
Comment (jdkent) [reply nimare/transforms.py:339]: what kind of metadata are you thinking about? Like whether the coordinates are "original" or generated through nimare?

Comment (tsalo) [reply nimare/transforms.py:339]: Exactly. It could have pretty much any reasonable name, since NIMADS doesn't have anything in it for this.

File: nimare/utils.py
Comment (jdkent) [reply nimare/utils.py:92]: yes, I believe so since when a non-essential column was added to `columns` (such as `z_stat`), the next time this comparison was performed, `columns` already contained `z_stat` and then the actual z_stat value for that study contrast was never added to the dataframe.

File: nimare/transforms.py
Comment (jdkent) [reply nimare/transforms.py:339]: Do you have a use case for a particular CBMA in mind?

File: nimare/utils.py
Comment (tsalo) [reply nimare/utils.py:92]: Thanks for catching it then. Could you just add a note about it to the PR summary?

File: nimare/transforms.py
Comment (jdkent) [reply nimare/transforms.py:373]: oops ,sorry I didn't see this, yeah that makes sense!

**Comment by jdkent** [comment]
several rebases later, I think I've covered @tsalo review comments, ready for another check before merging (make sure my rebases did not undo anything)

**Comment by jdkent** [comment]
@tsalo I think this is ready for merge (I just don't want to do it without your approval)

**Comment by tsalo** [comment]
Sorry yes I went through and resolve all of my outstanding comments. I was just planning to check things over one last time today, but honestly I'm happy to merge now and we can address any issues as they come up.

**Review by tsalo — APPROVED** [review]

**Pull request merged**
